nucypher/blockchain/eth/sol/source/contracts/WorkLock.sol

```diff
@@ -157,26 +211,164 @@ contract WorkLock {
     * @notice Cancel bid and refund deposited ETH
     */
     function cancelBid() external {
-        // TODO check date? check minimum amount of tokens? (#1508)
+        require(block.timestamp < endCancellationDate,
+            "Cancellation allowed only during cancellation window");
         WorkInfo storage info = workInfo[msg.sender];
         require(info.depositedETH > 0, "No bid to cancel");
         require(!info.claimed, "Tokens are already claimed");
         uint256 refundETH = info.depositedETH;
         info.depositedETH = 0;
-        if (block.timestamp <= endBidDate) {
-            ethSupply = ethSupply.sub(refundETH);
-        } else {
-            unclaimedTokens = unclaimedTokens.add(ethToTokens(refundETH));
+
+        // remove from bidders array, move last bidder to the empty place
+        uint256 length = bidders.length;
+        if (info.index != length - 1) {
+            address lastBidder = bidders[length - 1];
+            bidders[info.index] = lastBidder;
+            workInfo[lastBidder].index = info.index;
         }
+        bidders.pop();
+
+        ethSupply = ethSupply.sub(refundETH);
         msg.sender.sendValue(refundETH);
         emit Canceled(msg.sender, refundETH);
     }
 
+    /**
+    * @notice Cancels distribution, makes possible to retrieve all bids and creator gets all tokens
+    */
+    function shutdown() internal {
+        endCancellationDate = uint256(0) - 1; // "infinite" cancellation window
+        token.safeTransfer(creator, tokenSupply);
+    }
+
+    /**
+    * @notice Make force refund to bidders who can get tokens more than maximum allowed
+    * @param _biddersForRefund Sorted list of unique bidders. Only bidders who must receive a refund
```

Comment:
The most gas requiring part of this method is sending ETH back.
Also we can't trust to input data and need to find min value from stored data otherwise I think it could be used to decrease bid more than necessary